lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -67,22 +61,48 @@ class DeployTask
     # extensions/v1beta1/ReplicaSet -- managed by deployments
     # core/v1/Secret -- should not committed / managed by shipit
     def prune_whitelist
-      wl = %w(
-        core/v1/ConfigMap
-        core/v1/Pod
-        core/v1/Service
-        batch/v1/Job
-        extensions/v1beta1/DaemonSet
-        extensions/v1beta1/Deployment
-        apps/v1beta1/Deployment
-        extensions/v1beta1/Ingress
-        apps/v1beta1/StatefulSet
-        autoscaling/v1/HorizontalPodAutoscaler
-      )
-      if server_version >= Gem::Version.new('1.8.0')
-        wl << "batch/v1beta1/CronJob"
+      @prune_whitelist ||= _build_prune_whitelist
+    end
+
+    def _build_prune_whitelist
+      prunable_resources = all_resources.select(&:prunable?)
+      prunable_resources.map(&:qualified_kind)
+    end
+
+    def predeploy_sequence
+      @predeploy_sequence ||= _build_predeploy_sequence
+    end
+
+    def _build_predeploy_sequence
+      # Express dependencies as DAG
+      graph = RGL::DirectedAdjacencyGraph.new
+      graph.add_vertex(:ROOT_NODE)
+
+      # This is a partially ordered set, so we must make sure everything is reachable:
+      predeploy_resources = all_resources.select(&:predeploy?)
+      predeploy_resources.each { |res| graph.add_edge(:ROOT_NODE, res.kind) }
+
+      # Find resources that have explicit predeploy (inter-)dependencies:
+      predeploy_res_with_deps = all_resources.select(&:predeploy_dependencies)
+      predeploy_res_with_deps.each do |res|
+        # Edge [A,B] means B requires A to be deployed first
+        res.predeploy_dependencies.each { |dep| graph.add_edge(dep, res.kind) }
+      end
+
+      raise FatalDeploymentError, "Cyclic predeploy requirements: #{graph.cycles.flatten}" unless graph.cycles.empty?
+
+      # Topological sort is not unique, but will respect the requirements
+      predeploy_order = graph.topsort_iterator.to_a
+      predeploy_order.delete(:ROOT_NODE)
+      predeploy_order
+    end
+
+    def all_resources
+      resources = DiscoverableResource.all + KubernetesResource.all
+      # Omit unqualified kinds (they are unsupported on this cluster, or discovery hasn't been performed yet)
+      resources.select do |res|
+        res.constants.include?(:GROUP) && res.constants.include?(:VERSION)
```

Comment:
This is check is somewhat inelegant, but the `GROUP` and `VERSION` are now detected via discovery for most resources. If these fields are not present we can't query the API server, so we skip over them. This is not an issue if the user run discovery first.